community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureGDSFactory.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.proc;
+
+import org.neo4j.function.ThrowingFunction;
+import org.neo4j.graphdb.*;
+import org.neo4j.graphdb.event.KernelEventHandler;
+import org.neo4j.graphdb.event.TransactionEventHandler;
+import org.neo4j.graphdb.security.URLAccessRule;
+import org.neo4j.graphdb.security.URLAccessValidationError;
+import org.neo4j.kernel.api.KernelTransaction;
+import org.neo4j.kernel.api.ReadOperations;
+import org.neo4j.kernel.api.Statement;
+import org.neo4j.kernel.api.exceptions.ProcedureException;
+import org.neo4j.kernel.api.index.AutoIndexOperations;
+import org.neo4j.kernel.api.proc.CallableProcedure;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.api.index.AutoIndexing;
+import org.neo4j.kernel.impl.api.index.InternalAutoIndexing;
+import org.neo4j.kernel.impl.coreapi.CoreAPIAvailabilityGuard;
+import org.neo4j.kernel.impl.factory.GraphDatabaseFacade;
+import org.neo4j.kernel.impl.query.QueryExecutionEngine;
+import org.neo4j.kernel.impl.query.QueryExecutionKernelException;
+import org.neo4j.kernel.impl.query.QuerySession;
+import org.neo4j.kernel.impl.store.StoreId;
+
+import java.net.URL;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class ProcedureGDSFactory implements ThrowingFunction<CallableProcedure.Context,GraphDatabaseService,ProcedureException>
```

Comment:
Can't use that, 1) we have an argument per call to the factory and don't want to create it again and again (maybe we could do that) 2) we need to throw a checked exception, factory interface doesn't allow that